### PR TITLE
feat: cip64 and eip1559

### LINKED ISSUE #11: Add Support for CIp64 Transactions
Current

celo-ethers-wrapper currently only supports legacy style transaction encoding. Supporting gasPrice, feeCurrency and GateWay fields. However gateway is deprecated. 

Future

support [cip64 ](https://github.com/celo-org/celo-proposals/blob/master/CIPs/cip-0064.md) which can be thought of as FeeMarket eip1559 tx + feeCurrency field.

### PR DESCRIPTION
**Pull request by nicolasbrugneaux**
This PR adds supports for CIP64 and EIP1559.

[Closes #11 ]

### CODE REVIEW

**Comment by nicolasbrugneaux** [comment]
Notes before the weekend: it should be working but it isn't? (at least celo-legacy isn't broken)
Getting this error for now
```
Sending 1 CELO wei with cUSD feeCurrency
{
  type: 123,
  chainId: 44787,
  nonce: 194,
  maxPriorityFeePerGas: '0x012a05f200',
  maxFeePerGas: '0x012a05f200',
  gasLimit: BigNumber { _hex: '0x6aa4', _isBigNumber: true },
  to: '0x21eCC6113BfA08c428cA11baE55Ba038FcA0e7Da',
  value: BigNumber { _hex: '0x01', _isBigNumber: true },
  data: '0x',
  feeCurrency: '0x874069Fa1Eb16D44d622F2e0Ca25eeA172369bC1',
  v: 89609,
  r: '0xc8d84204882131776a3a2fe0cfd8c837cab6bfe2ede9e280706158f755c40a5f',
  s: '0x3439c9cbd8e2b19ec2125645061325d6fc39ce90c1072b66af206d13aea31fc0',
  from: '0x21eCC6113BfA08c428cA11baE55Ba038FcA0e7Da',
  hash: '0xdbd7d8e707b1354ad2e7542a6ac49d4c848f1695564042d755602f16f27bb813'
}
Error: processing response error (body="{\"jsonrpc\":\"2.0\",\"id\":74,\"error\":{\"code\":-32000,\"message\":\"invalid sender\"}}\n", error={"code":-32000}, requestBody="{\"method\":\"eth_sendRawTransaction\",\"params\":[\"0x7bf88782aef381c285012a05f20085012a05f200826aa49421ecc6113bfa08c428ca11bae55ba038fca0e7da0180c094874069fa1eb16d44d622f2e0ca25eea172369bc183015e09a0c8d84204882131776a3a2fe0cfd8c837cab6bfe2ede9e280706158f755c40a5fa03439c9cbd8e2b19ec2125645061325d6fc39ce90c1072b66af206d13aea31fc0\"],\"id\":74,\"jsonrpc\":\"2.0\"}", requestMethod="POST", url="https://alfajores-forno.celo-testnet.org", code=SERVER_ERROR, version=web/5.7.1)
    at Logger.makeError (/Users/nicolasbrugneaux/workspace/celo-ethers-wrapper/node_modules/@ethersproject/logger/src.ts/index.ts:269:28)
    at Logger.throwError (/Users/nicolasbrugneaux/workspace/celo-ethers-wrapper/node_modules/@ethersproject/logger/src.ts/index.ts:281:20)
    at /Users/nicolasbrugneaux/workspace/celo-ethers-wrapper/node_modules/@ethersproject/web/src.ts/index.ts:341:28
    at step (/Users/nicolasbrugneaux/workspace/celo-ethers-wrapper/node_modules/@ethersproject/web/lib/index.js:33:23)
    at Object.next (/Users/nicolasbrugneaux/workspace/celo-ethers-wrapper/node_modules/@ethersproject/web/lib/index.js:14:53)
    at fulfilled (/Users/nicolasbrugneaux/workspace/celo-ethers-wrapper/node_modules/@ethersproject/web/lib/index.js:5:58)
    at processTicksAndRejections (node:internal/process/task_queues:95:5) {
  reason: 'processing response error',
  code: 'SERVER_ERROR',
  body: '{"jsonrpc":"2.0","id":74,"error":{"code":-32000,"message":"invalid sender"}}\n',
  error: Error: invalid sender
      at getResult (/Users/nicolasbrugneaux/workspace/celo-ethers-wrapper/node_modules/@ethersproject/providers/src.ts/json-rpc-provider.ts:142:28)
      at processJsonFunc (/Users/nicolasbrugneaux/workspace/celo-ethers-wrapper/node_modules/@ethersproject/web/src.ts/index.ts:383:22)
      at /Users/nicolasbrugneaux/workspace/celo-ethers-wrapper/node_modules/@ethersproject/web/src.ts/index.ts:320:42
      at step (/Users/nicolasbrugneaux/workspace/celo-ethers-wrapper/node_modules/@ethersproject/web/lib/index.js:33:23)
      at Object.next (/Users/nicolasbrugneaux/workspace/celo-ethers-wrapper/node_modules/@ethersproject/web/lib/index.js:14:53)
      at fulfilled (/Users/nicolasbrugneaux/workspace/celo-ethers-wrapper/node_modules/@ethersproject/web/lib/index.js:5:58)
      at processTicksAndRejections (node:internal/process/task_queues:95:5) {
    code: -32000,
    data: undefined
  },
  requestBody: '{"method":"eth_sendRawTransaction","params":["0x7bf88782aef381c285012a05f20085012a05f200826aa49421ecc6113bfa08c428ca11bae55ba038fca0e7da0180c094874069fa1eb16d44d622f2e0ca25eea172369bc183015e09a0c8d84204882131776a3a2fe0cfd8c837cab6bfe2ede9e280706158f755c40a5fa03439c9cbd8e2b19ec2125645061325d6fc39ce90c1072b66af206d13aea31fc0"],"id":74,"jsonrpc":"2.0"}',
  requestMethod: 'POST',
  url: 'https://alfajores-forno.celo-testnet.org',
  transaction: {
    type: 123,
    chainId: 44787,
    nonce: 194,
    maxPriorityFeePerGas: '0x012a05f200',
    maxFeePerGas: '0x012a05f200',
    gasLimit: BigNumber { _hex: '0x6aa4', _isBigNumber: true },
    to: '0x21eCC6113BfA08c428cA11baE55Ba038FcA0e7Da',
    value: BigNumber { _hex: '0x01', _isBigNumber: true },
    data: '0x',
    feeCurrency: '0x874069Fa1Eb16D44d622F2e0Ca25eeA172369bC1',
    v: 89609,
    r: '0xc8d84204882131776a3a2fe0cfd8c837cab6bfe2ede9e280706158f755c40a5f',
    s: '0x3439c9cbd8e2b19ec2125645061325d6fc39ce90c1072b66af206d13aea31fc0',
    from: '0x21eCC6113BfA08c428cA11baE55Ba038FcA0e7Da',
    hash: '0xdbd7d8e707b1354ad2e7542a6ac49d4c848f1695564042d755602f16f27bb813'
  },
  transactionHash: '0xdbd7d8e707b1354ad2e7542a6ac49d4c848f1695564042d755602f16f27bb813'
}
```


Could use some advice here @jmrossy

File: src/lib/CeloProvider.ts [hunk 1]
```diff
@@ -33,17 +13,21 @@ export class CeloProvider extends providers.JsonRpcProvider {
...
         signedTransaction: hexTx,
       });
-      return this._wrapTransaction(tx, hash);
-    } catch (error: any) {
-      error.transaction = tx;
-      error.transactionHash = tx.hash;
+      return this._wrapTransaction(tx, hash, blockNumber);
+    } catch (error) {
+      (<any>error).transaction = tx;
+      (<any>error).transactionHash = tx.hash;
```
Comment (nicolasbrugneaux) [inline src/lib/CeloProvider.ts:30]: backported from ethers v5

**Comment by jmrossy** [comment]
@nicolasbrugneaux The `invalid sender` error makes me think the tx is missing a `from` field but I see there is one in your example transaction.

My recommendation would be to upgrade to the latest `ethers` v5 package version (5.7.2) and then try again

File: src/lib/transactions.ts [hunk 2]
```diff
@@ -463,30 +482,43 @@ export function parseCeloTransaction(
...
+    // // NOTE: is this condition even correct now?
+    // if (!tx.chainId) {
+    //   const chainId = Math.max(0, Math.floor((tx.v - 35) / 2));
+    //   tx.chainId = chainId;
+    // }
+    // let recoveryParam = tx.v - 27;
+    // if (tx.chainId !== 0) {
+    //   // NOTE: inverse operation that was done in `serializeCeloTransaction`
+    //   recoveryParam -= tx.chainId * 2 + 8;
+    // }
```
Comment (jmrossy) [inline src/lib/transactions.ts:514]: It might not be correct anymore, ethers.js changed/reorganized this code after I copied it in here a while ago

**Comment by nicolasbrugneaux** [comment]
Ready for review! :)

File: src/lib/CeloWallet.ts [hunk 3]
```diff
@@ -111,11 +122,12 @@ export class CeloWallet extends Wallet {
    * https://github.com/ethers-io/ethers.js/blob/master/packages/abstract-signer/src.ts/index.ts
    */
   async estimateGas(
-    transaction: utils.Deferrable<CeloTransactionRequest>
+    transaction: utils.Deferrable<CeloTransaction>
   ): Promise<BigNumber> {
     this._checkProvider("estimateGas");
-    const tx = await utils.resolveProperties(transaction);
-    return await this.provider.estimateGas(tx);
+    // @ts-expect-error
```
Comment (jmrossy) [inline src/lib/CeloWallet.ts:128]: In my experience, ts-expect-error can sometimes cause errors to be raised downstream in typescript packages that use this package. I'd rather we use `transaction as any` instead.

Comment (jmrossy) [inline src/lib/CeloWallet.ts:128]: In my experience, ts-expect-error can sometimes cause errors to be raised downstream in typescript packages that use this package. I'd rather we use transaction as any instead.

File: src/lib/transaction/utils.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  return (
+    value === undefined ||
+    value === null ||
+    value === "0" ||
+    (typeof value == "string"
+      ? value.toLowerCase() === "0x" || value.toLowerCase() === "0x0"
+      : hexlify(value) === "0x0")
+  );
+}
+function isPresent(value: string | BigNumberish | undefined | null) {
```
Comment (jmrossy) [inline src/lib/transaction/utils.ts:14]: Nit: spacing above

File: src/lib/transaction/utils.ts [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    .filter((key) => !keys.includes(key as keyof T))
+    .reduce((acc, key) => {
+      acc[key as keyof T] = object[key as keyof T];
+      return acc;
+    }, {} as T);
+}
+
+export function adjustForGasInflation(gas: BigNumber): BigNumber {
+  // NOTE: Don't ask me
+  const GAS_INFLATION_FACTOR = 1.3;
```
Comment (jmrossy) [inline src/lib/transaction/utils.ts:59]: Please move this to a new file called `consts.ts` and include a more detailed comment

**Review by jmrossy — COMMENTED** [review]
Nice work! Just a few minor things to change but overall looks good

**Review by jmrossy — APPROVED** [review]

**Comment by jmrossy** [comment]
Okay looks good. Are there any other changes/improvements you'd like to make before merging?

**Comment by nicolasbrugneaux** [comment]
I can seem to merge this despite your approval. Probably don't have the rights.

**Comment by jmrossy** [comment]
Merged and published as 1.0.0. Thanks for the contribution!

**Pull request merged**